File: 238. Product of Array Except Self/solution.py
from typing import List


class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        product = 1
        products = [1] * len(nums)

        for i in range(len(nums)):
            products[i] *= product
            product *= nums[i]

        product = 1

        for i in range(len(nums) - 1, -1, -1):
            products[i] *= product
            product *= nums[i]

        return products

File: 238. Product of Array Except Self/Solution.cs
public class Solution {
    public int[] ProductExceptSelf(int[] nums) {
        int product = 1;
        int[] products = new int[nums.Length];
        Array.Fill(products, 1);

        for (int i = 0; i < nums.Length; i++) {
            products[i] = product;
            product *= nums[i];
        }

        product = 1;

        for (int i = nums.Length - 1; i >= 0; i--) {
            products[i] *= product;
            product *= nums[i];
        }

        return products;
    }
}

// Time complexity: O(n)
// Space complexity: O(1)